Stripe/STPAPIClient+ApplePay.swift

```diff
@@ -0,0 +1,192 @@
+  /// - Parameter stripeError:   An error from the Stripe SDK.
+  public class func pkPaymentError(forStripeError stripeError: Error?) -> Error? {
+    guard let stripeError = stripeError else {
+      return nil
+    }
+
+    if (stripeError as NSError).domain == STPError.stripeDomain
+      && ((stripeError as NSError).userInfo[STPError.cardErrorCodeKey] as? String
+        == STPCardErrorCode.incorrectZip.rawValue)
+    {
+      var userInfo = (stripeError as NSError).userInfo
+      var errorCode: PKPaymentError.Code = .unknownError
+      errorCode = .billingContactInvalidError
+      userInfo[PKPaymentErrorKey.postalAddressUserInfoKey.rawValue] = CNPostalAddressPostalCodeKey
+      return NSError(domain: STPError.stripeDomain, code: errorCode.rawValue, userInfo: userInfo)
+    }
+    return stripeError
+  }
+
+  class func billingDetails(from payment: PKPayment) -> STPPaymentMethodBillingDetails? {
+    var billingDetails: STPPaymentMethodBillingDetails?
+    if payment.billingContact != nil {
+      billingDetails = STPPaymentMethodBillingDetails()
+      var billingAddress: STPAddress?
+      if let billingContact = payment.billingContact {
+        billingAddress = STPAddress(pkContact: billingContact)
+      }
+      billingDetails?.name = billingAddress?.name
+      billingDetails?.email = billingAddress?.email
+      billingDetails?.phone = billingAddress?.phone
+      if payment.billingContact?.postalAddress != nil {
+        if let billingAddress = billingAddress {
+          billingDetails?.address = STPPaymentMethodAddress(address: billingAddress)
+        }
+      }
+    }
+
+    // The phone number and email in the "Contact" panel in the Apple Pay dialog go into the shippingContact,
+    // not the billingContact. To work around this, we should fill the billingDetails' email and phone
+    // number from the shippingDetails.
+    if payment.shippingContact != nil {
+      var shippingAddress: STPAddress?
+      if let shippingContact = payment.shippingContact {
+        shippingAddress = STPAddress(pkContact: shippingContact)
+      }
+      if billingDetails?.email == nil && shippingAddress?.email != nil {
+        if billingDetails == nil {
+          billingDetails = STPPaymentMethodBillingDetails()
+        }
+        billingDetails?.email = shippingAddress?.email
+      }
+      if billingDetails?.phone == nil && shippingAddress?.phone != nil {
+        if billingDetails == nil {
+          billingDetails = STPPaymentMethodBillingDetails()
+        }
+        billingDetails?.phone = shippingAddress?.phone
+      }
+    }
+
+    return billingDetails
+  }
+
+  class func addressParams(from contact: PKContact?) -> [AnyHashable: Any]? {
+    guard let contact = contact else {
+      return nil
+    }
+    var params: [AnyHashable: Any] = [:]
+    let stpAddress = STPAddress(pkContact: contact)
+
+    params["name"] = stpAddress.name
+    params["address_line1"] = stpAddress.line1 ?? ""
```

Comment:
I don't think we should add empty values here

Reply:
++